src/decoder/plugins/VgmstreamDecoderPlugin.cxx

```diff
@@ -0,0 +1,332 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+// Copyright The Music Player Daemon Project
+
+#include "VgmstreamDecoderPlugin.hxx"
+#include "../DecoderAPI.hxx"
+#include "Log.hxx"
+#include "fs/AllocatedPath.hxx"
+#include "fs/NarrowPath.hxx"
+#include "fs/Path.hxx"
+#include "lib/fmt/PathFormatter.hxx"
+#include "pcm/CheckAudioFormat.hxx"
+#include "song/DetachedSong.hxx"
+#include "tag/Builder.hxx"
+#include "tag/Handler.hxx"
+#include "tag/Names.hxx"
+#include "tag/ReplayGainParser.hxx"
+#include "util/Domain.hxx"
+#include "util/ScopeExit.hxx"
+#include "util/StringCompare.hxx"
+
+extern "C" {
+#include <vgmstream/libvgmstream.h>
+}
+
+#include <cassert>
+
+#define SUBSONG_PREFIX "song_"
+
+struct VgmstreamContainerPath {
+	AllocatedPath path;
+	int subsong;
+};
+
+static constexpr Domain vgmstream_domain("vgmstream");
+
+static libvgmstream_config_t cfg = {};
+
+static void
+VgmstreamGetDouble(const ConfigBlock &block, double *out, const char *name,
+		   double default_value)
+{
+	if (const BlockParam *param = block.GetBlockParam(name))
+		*out = std::stod(param->value);
+	else
+		*out = default_value;
+}
+
+static bool
+VgmstreamInit(const ConfigBlock &block)
+{
+	const uint32_t version = libvgmstream_get_version();
+	unsigned major = version >> 24;
+	unsigned minor = version >> 16 & 0xFF;
+	unsigned patch = version & 0xFFFF;
+	FmtDebug(vgmstream_domain, "vgmstream {}.{}.{}", major, minor, patch);
+
+	cfg.ignore_loop = block.GetBlockValue("ignore_loop", false);
+	cfg.force_loop = block.GetBlockValue("force_loop", false);
+	cfg.really_force_loop = block.GetBlockValue("really_force_loop", false);
+	cfg.ignore_fade = block.GetBlockValue("ignore_fade", false);
+
+	// same defaults as plugins shipped with vgmstream
+	VgmstreamGetDouble(block, &cfg.loop_count, "loop_count", 2.0);
+	VgmstreamGetDouble(block, &cfg.fade_time, "fade_time", 10.0);
+	VgmstreamGetDouble(block, &cfg.fade_delay, "fade_delay", 0.0);
+
+	return true;
+}
+
+static AudioFormat
+VgmstreamGetFormat(libvgmstream_t *lib, libstreamfile_t *file)
+{
+	uint32_t sample_rate = lib->format->sample_rate;
+	auto sample_format = SampleFormat::UNDEFINED;
+	uint8_t channels = lib->format->channels;
+
+	if (lib->format->sample_format == LIBVGMSTREAM_SFMT_PCM16) {
+		sample_format = SampleFormat::S16;
+	} else if (lib->format->sample_format == LIBVGMSTREAM_SFMT_PCM24) {
+		// mpd doesn't support S24 samples, so reopen the stream as S32
```

Comment:
Oh yes, MPD does.
https://github.com/MusicPlayerDaemon/MPD/blob/f43d7de99f84f32099f0c9ef92ea96e2306a9d13/src/pcm/SampleFormat.hxx#L22-L26

Reply:
From what I can tell, vgmstream's PCM24 format is *packed* 24-bit samples. This seemed easier than writing code to pad to 32 bits, though I can try doing that instead.

Reply:
Yes, please. There is `pcm_unpack_24()` which does this for you.